src/ast/element-descriptor.ts

```diff
@@ -12,72 +12,36 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-import * as dom5 from 'dom5';
-import {VisitorOption, traverse} from 'estraverse';
 import * as estree from 'estree';
 
 import {SourceLocation} from '../elements-format';
-import {VisitResult, Visitor} from '../javascript/estree-visitor';
 import * as jsdoc from '../javascript/jsdoc';
-import {Document} from '../parser/document';
 
-import {BehaviorOrName, Descriptor, EventDescriptor, LiteralValue, LocationOffset, PropertyDescriptor, correctSourceLocation} from './ast';
+import {Document, Event, Feature, LocationOffset, Property, Resolvable, ScannedEvent, ScannedFeature, ScannedProperty, correctSourceLocation} from './ast';
 
 export {Visitor} from '../javascript/estree-visitor';
 
-export interface Options {
-  type: 'element'|'behavior';
-  is?: string;
-
-  jsdoc?: jsdoc.Annotation;
-  desc?: string;
-  contentHref?: string;
-  properties?: PropertyDescriptor[];
-  observers?: {
-    javascriptNode: estree.Expression | estree.SpreadElement,
-    expression: LiteralValue
-  }[];
-  behaviors?: BehaviorOrName[];
-
-  demos?: {desc: string; path: string}[];
-  events?: EventDescriptor[];
-  hero?: string;
-  domModule?: dom5.Node;
-  scriptElement?: dom5.Node;
-
-  abstract?: boolean;
-  sourceLocation?: SourceLocation;
+export interface ScannedAttribute {
+  name: string;
+  sourceLocation: SourceLocation;
+  description?: string;
+  type?: string;
 }
 
-/**
- * The metadata for a single polymer element
- */
-export class ElementDescriptor implements Descriptor {
-  type: 'element'|'behavior';
-  is: string;
+export class ScannedElement implements ScannedFeature, Resolvable {
+  tagName?: string;
+  className?: string;
+  superClass?: string;
+  extends?: string;
+  properties: ScannedProperty[] = [];
+  attributes: ScannedAttribute[] = [];
+  description = '';
+  demos: {desc?: string; path: string}[] = [];
+  events: ScannedEvent[] = [];
+  sourceLocation: SourceLocation;
+  node: estree.Node;
 
   jsdoc?: jsdoc.Annotation;
-  desc?: string;
-  contentHref?: string;
-  properties: PropertyDescriptor[] = [];
-  observers: {
-    javascriptNode: estree.Expression | estree.SpreadElement,
-    expression: LiteralValue
-  }[] = [];
-  behaviors: BehaviorOrName[] = [];
-
-  demos: {desc: string; path: string}[] = [];
-  events: EventDescriptor[] = [];
-  hero?: string;
-  domModule?: dom5.Node;
-  scriptElement?: dom5.Node;
-
-  abstract?: boolean;
-  sourceLocation?: SourceLocation;
-
-  constructor(options: Options) {
-    Object.assign(this, options);
-  }
 
   applyLocationOffset(locationOffset?: LocationOffset) {
```

Comment:
can we refactor this so that `sourceLocation` is set to the correct value at scan time?